### PR TITLE
Bridged Device Basic Information cluster

### PR DESCRIPTION
**Pull request by lzgrablic02**
#### Problem
* Spec conformance.
* ReachableChanged implementation.

#### Change overview
XML and ZAP updates.
ReachableChanged implementation.

#### Testing
How was this tested? (at least one bullet point required)
* By temporarily re-adding in all-clusters-app and making Reachable writeable:
  * Saw ReachableChanged event in log when writing to Reachable.
  * No event when writing to NodeLabel.

### CODE REVIEW

**Comment by lzgrablic02** [comment]
Not sure why this built locally with the files misnamed...

File: src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+    CHIP_ERROR status = CHIP_NO_ERROR;
+
+    switch (aPath.mAttributeId)
+    {
+    case VendorName::Id: {
+        constexpr size_t kMaxLen     = DeviceLayer::ConfigurationManager::kMaxVendorNameLength;
+        char vendorName[kMaxLen + 1] = { 0 };
+        status                       =
+            DeviceLayer::PlatformMgr().GetBridgedVendorName(aPath.mEndpointId, vendorName, sizeof(vendorName));
```
Comment (bzbarsky-apple) [inline src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp:74]: It doesn't really make sense to ask the platform manager for this info.

On an actual bridge, the way this would work is that the endpoint would be dynamic and the bridge itself would implement the attributes either via AttributeAccessInterface or via just marking them external and handling the callbacks.  I don't think it makes any sense to have a "non-bridge" implementation of this cluster for all-clusters-app...

Comment (lzgrablic02) [reply src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp:74]: Is there documentation on the architecture of attribute access? I'm not sure if I fully understand it. Would handling it via AttributeAccessInterface just mean using the generated accessors? And how would it handle the callbacks differently if they stay marked external?

You mean removing it entirely from all-clusters-app? Is there a reason it was already in there, as well as some other clusters?

Comment (bzbarsky-apple) [reply src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp:74]: There is some documentation at https://github.com/project-chip/connectedhomeip/blob/master/src/app/AttributeAccessInterface.h#L367-L385 but not anything that describes the overall set of options for attribute storage...  I'll see if I can write up something.

But in brief, there are four options:

1. Attribute handled entirely by the managed attribute store.  Reads and writes are handled automatically, the attribute is stored in RAM and will also be written through to NVM if the "NVM" option is selected in ZAP.  Writes to the attribute call notification functions that the cluster and the application can implement to find out when the value of the attribute changed.  This option is only available for static endpoints that the code generator knows about, so can allocate space for them in the attribute store.
2. Attribute handled by an AttributeAccessInterface implemented in the SDK itself.  This is used to handle attributes where we want to run some logic on every read, or for attributes that the generic attribute store does not know how to store (lists and structs).  If the attribute is not marked as "External" storage, this can fall back to option 1 as described at the link above.
3. Attribute handled by an AttributeAccessInterface implemented in the application.  This makes sense when the SDK itself could not very easily provide the values.  This situation could _also_ be handled by doing option 2 and having the SDK implementation talk to the application via some defined delegate interface, which is more or less what this PR is doing, right?  But if the only logic in the SDK is glue code, why bother?
4. Attribute handled by application external attribute callbacks.  These get called for attributes with "External" storage that are not handled by an AttributeAccessInterface.  It's not clear to me that we really need this option, since all consumers of it can in fact use an AttributeAccessInterface, but so far we have not gotten rid of it (it came with the original Silicon Labs data model code).

I assume this cluster is in all-clusters-app because (1) someone took the name literally and (2) people wanted to run YAML tests against this cluster.  But since this cluster really only makes sense for dynamic endpoints (bridged ones), what we should really do is remove it from all-clusters-app, implement in some sane way in bridge-app, and make sure that we can run some subset of the YAML tests against bridge-app.

Comment (lzgrablic02) [reply src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp:74]: Thanks for the clarifications!

Comment (yufengwangca) [reply src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp]: @bzbarsky-apple 

I agree VendorName attribute should be implemented on bridge device instead of the node communicating with a Bridge(bridged device), shall we review the following attributes defined in the spec for the node communicating with a Bridge.

DataModelRevision | X
VendorName | O
VendorID | O
ProductName | O
ProductID | X
NodeLabel | O
Location | X
HardwareVersion | O
HardwareVersionString | O
SoftwareVersion | O
SoftwareVersionString | O
ManufacturingDate | O
PartNumber | O
ProductURL | O
ProductLabel | O
SerialNumber | O
LocalConfigDisabled | X
Reachable | M
UniqueID | O

The spec does not say which attributes should reside on the bridge device or the bridged device itself

Comment (yufengwangca) [reply src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp]: We need to review which attributes shall be implemented on the bridge device or the bridged device (Node communicating with the bridge), the spec does not say it clearly.

DataModelRevision | X
VendorName | O
VendorID | O
ProductName | O
ProductID | X
NodeLabel | O
Location | X
HardwareVersion | O
HardwareVersionString | O
SoftwareVersion | O
SoftwareVersionString | O
ManufacturingDate | O
PartNumber | O
ProductURL | O
ProductLabel | O
SerialNumber | O
LocalConfigDisabled | X
Reachable | M
UniqueID | O

Comment (bzbarsky-apple) [reply src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp]: The bridge needs to implement these attributes (modulo optionality).  Note that the only mandatory attribute here is "Reachable".

Whether it implements them by querying the bridged device every time, or by caching information about it, or by having the user manually input the information once and saving it is up to the bridge app implementation.

And the one required thing (Reachable) clearly needs to be implemented by the bridge itself.

Comment (lzgrablic02) [reply src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp]: I was intending to submit this in a separate PR. Can this PR go in as-is?

File: examples/all-clusters-app/mbed/CMakeLists.txt [hunk 2]
```diff
@@ -90,7 +90,6 @@ target_sources(${APP_TARGET} PRIVATE
                ${APP_CLUSTERS}/application-launcher-server/application-launcher-server.cpp
                ${APP_CLUSTERS}/audio-output-server/audio-output-server.cpp   
                ${APP_CLUSTERS}/barrier-control-server/barrier-control-server.cpp
-               ${APP_CLUSTERS}/basic/basic.cpp
```
Comment (bzbarsky-apple) [inline examples/all-clusters-app/mbed/CMakeLists.txt:93]: Why this change?

File: src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    EventNumber eventNumber;
+    if (CHIP_NO_ERROR != LogEvent(event, endpointId, eventNumber))
+    {
+        ChipLogError(Zcl, "ReachableChanged: Failed to record ReachableChanged event");
+    }
+}
+
+} // anonymous namespace
+
+void MatterBridgedDeviceBasicClusterServerAttributeChangedCallback(const ConcreteAttributePath & attributePath)
```
Comment (bzbarsky-apple) [inline src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp:53]: So this is envisaging the bridge writing the "reachable" attribute when it determines that it's changed, and that landing in this code?

I guess that works, ok.

**Review by bzbarsky-apple — APPROVED** [review]
Looks like this got rescoped from the last time I looked at it... it's ok as far as it goes, though I am not sure the indirect "write the attribute that you actually implement yourself" thing is really the right API here....

File: examples/all-clusters-app/mbed/CMakeLists.txt (hunk 2 above)
Comment (lzgrablic02) [reply examples/all-clusters-app/mbed/CMakeLists.txt:93]: Duplicate line.

File: src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp (hunk 3 above)
Comment (lzgrablic02) [reply src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp:53]: It'll trigger the event any time `Reachable` is written to.

**Review by harsha-rajendran — APPROVED** [review]
